### PR TITLE
Alpha node swap support

### LINKED ISSUE #53533: Kubelet/Kubernetes should work with Swap Enabled
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:

Kubelet/Kubernetes 1.8 does not work with Swap enabled on Linux Machines.

I have found this original issue https://github.com/kubernetes/kubernetes/issues/31676
This PR https://github.com/kubernetes/kubernetes/pull/31996 
and last change which enabled it by default https://github.com/kubernetes/kubernetes/commit/71e8c8eba43a0fade6e4edfc739b331ba3cc658a

If Kubernetes does not know how to handle memory eviction when Swap is enabled - it should find a way how to do that, but not asking to get rid of swap.

Please follow kernel.org [Chapter 11  Swap Management](https://www.kernel.org/doc/gorman/html/understand/understand014.html), for example

> The casual reader may think that with a sufficient amount of memory, swap is unnecessary but this brings us to the second reason. A significant number of the pages referenced by a process early in its life may only be used for initialisation and then never used again. It is better to swap out those pages and create more disk buffers than leave them resident and unused.

In case of running a lot of node/java applications I have seen always a lot of pages are swapped, just because they aren't used anymore.

**What you expected to happen**:

Kubelet/Kubernetes should work with Swap enabled. I believe instead of disabling swap and giving users no choices kubernetes should support more use cases and various workloads, some of them can be an applications which might rely on caches.

I am not sure how kubernetes decided what to kill with memory eviction, but considering that Linux has this capability, maybe it should align with how Linux does that? https://www.kernel.org/doc/gorman/html/understand/understand016.html

I would suggest to rollback the change for failing when swap is enabled, and revisit how the memory eviction works currently in kubernetes. Swap can be important for some workloads. 

**How to reproduce it (as minimally and precisely as possible)**:

Run kubernetes/kublet with default settings on linux box

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/sig node
cc @mtaufen @vishh @derekwaynecarr @dims

### PR DESCRIPTION
**Pull request by ehashman**
#### What type of PR is this?

/kind feature
/kind api-change
/sig node

#### What this PR does / why we need it:

Adds swap support per KEP-2400.

#### Which issue(s) this PR fixes:

Fixes #53533.

#### Special notes for your reviewer:

Design details in https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#design-details

#### Does this PR introduce a user-facing change?

```release-note
Alpha swap support can now be enabled on Kubernetes nodes with the NodeSwapEnabled feature flag. See <website link> for details.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/
```

### CODE REVIEW

**Comment by ehashman** [comment]
Ah, still waiting on https://github.com/kubernetes/test-infra/pull/22453/files to merge.

**Comment by ehashman** [comment]
/triage accepted

This should be ready for a round of initial review now.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (ehashman) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:102]: So this change is deliberate, to make the CRI behave like dockershim:

https://github.com/kubernetes/kubernetes/blob/24b716673caae31f070b06a337bc12c97ff1d4cb/pkg/kubelet/dockershim/helpers_linux.go#L115-L117

And yeah, the assumption is that there won't be any significant swap usage with these controls set.

[From the KEP:](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#just-set---fail-swap-onfalse)

> Dockershim currently sets swap available for workloads to 0. The CRI does not restrict it at all. This inconsistency makes it difficult or impossible to use swap in production, particularly if a user wants to restrict workloads from using swap when using the CRI rather than dockershim.

I could update the KEP to use a different name if you think "NoSwap" is misleading and we get general agreement. I can also improve the documentation to match.

**Comment by ehashman** [comment]
/test pull-kubernetes-node-swap-fedora

need to fix that name, haha

**Comment by ehashman** [comment]
Ubuntu test passed but there are no logs, want to try again to see if it's an issue with the test job or just that run /test pull-kubernetes-node-kubelet-swap-ubuntu

**Comment by ehashman** [comment]
Rebased for CRI changes that landed ahead of this one.

/test pull-kubernetes-node-swap-fedora

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (ehashman) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:104]: (name change is done)

Comment (ehashman) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:104]: KEP update is https://github.com/kubernetes/enhancements/pull/2794

**Comment by ehashman** [comment]
/label api-review

for the kubelet config changes

/skip

File: pkg/kubelet/apis/config/types.go
Comment (liggitt) [inline pkg/kubelet/apis/config/types.go:576]: I expected to see something in here indicating whether swap support was enabled, defaulting to the current behavior of the kubelet not allowing swap. Is this feature intended to unconditionally support swap once it graduates to beta/GA?

File: pkg/kubelet/apis/config/validation/validation.go
Comment (liggitt) [inline pkg/kubelet/apis/config/validation/validation.go:162]: if the feature gate is not enabled, should trying to set config fields for the feature produce a validation error?

File: pkg/kubelet/apis/config/types.go
Comment (ehashman) [reply pkg/kubelet/apis/config/types.go:576]: In order to enable swap, one must also enable the `--fail-swap-on=false` flag on the kubelet. Otherwise, kubelet will not start when there is swap on the system.

So matrix is:

| |fail-swap-on=true|fail-swap-on=false|
|-|-|-|
|feature-flag=false|Kubelet won't start if swap is available|Kubelet will default to current behaviour|
|feature-flag=true|Kubelet won't start if swap is available|Kubelet will use swap config|

My plan is to eventually remove the feature flag, but there are no plans for removing the `--fail-swap-on` flag.

File: pkg/kubelet/apis/config/validation/validation.go
Comment (ehashman) [reply pkg/kubelet/apis/config/validation/validation.go:162]: I'm not sure of the API convention here, should it?

File: pkg/kubelet/apis/config/types.go
Comment (liggitt) [reply pkg/kubelet/apis/config/types.go:576]: ah, I missed the separate `failSwapOn` field in the kubelet config file. that's what I was looking for

File: pkg/kubelet/apis/config/validation/validation.go
Comment (liggitt) [reply pkg/kubelet/apis/config/validation/validation.go:162]: for config, we try to fail fast/loudly when someone passes flags or configuration that is ineffective or doesn't make sense

Comment (ehashman) [reply pkg/kubelet/apis/config/validation/validation.go:162]: I see CPU Manager doing this as well, will add validation.

Comment (ehashman) [reply pkg/kubelet/apis/config/validation/validation.go:162]: e8afdbfceb23398fccf8665535b133ddd118f87d

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (liggitt) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:965]: move the comments about allowed values to be field doc on the `SwapBehavior` field (and start with `swapBehavior` to match the serialized field name (and the update being made in https://github.com/kubernetes/kubernetes/pull/96699)

File: pkg/kubelet/apis/config/validation/validation.go
Comment (liggitt) [inline pkg/kubelet/apis/config/validation/validation.go:163]: if setting any field under MemorySwap would not make sense if NodeSwapEnabled is false, compare the whole struct so any future fields are caught as well (`kc.MemorySwap != (kubeletconfig.MemorySwapConfiguration{})`

**Comment by liggitt** [comment]
API change lgtm, just a couple nits on the doc and future-proofing the validation check

**Comment by ehashman** [comment]
Rebased and all API review comments should be addressed.

/test pull-kubernetes-node-kubelet-swap-ubuntu
/test pull-kubernetes-node-swap-fedora

**Comment by liggitt** [comment]
API bits lgtm, ping once the functional bits have lgtm from node reviewers and I can add the approved tag

**Comment by ehashman** [comment]
flake #99230 
/retest

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (derekwaynecarr) [inline pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:99]: just recording this as this swaps out of my own mental cache sometimes.

if a container has a defined memory limit X, it will still have `MemoryLimitInBytes=X`, but it may now use unbounded additional swap by setting `MemorySwapLimitInBytes` if `UnlimitedSwap` is enabled.  This is consistent with existing behavior where --fail-swap-on was false because no kubelet enforced limit was written.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (derekwaynecarr) [inline pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:100]: just a minor nit, but can we include the token for case: LimitedSwap and have it passthrough to default: so its clear to future authors/reviewers in this space?

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (derekwaynecarr) [inline pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:93]: I think we need a similar change in `ResourceConfigForPod` for pod level cgroup settings created by pod cgroup manager.  I would expect them to match the container settings.  I _think_ memory backed volumes could ultimately use swap, but would like @sjenning to confirm.  Either way, the cgroup settings for memory should match pod and container scopes.

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
Two updates requested:

- update pod level cgroup creation to match container cgroup creation
- minor nit on the case stmt to enumerate all options for readability

Thanks!

**Comment by derekwaynecarr** [comment]
@ehashman We can skip the `ResourceConfigForPod` change until Beta per the KEP https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#beta

The default value for pod cgroup will basically be unbounded in practice, and it is not until the beta criteria that its value conceptually could vary further than unbounded.

**Comment by derekwaynecarr** [comment]
thanks for updates.

/lgtm
/approve

**Comment by ehashman** [comment]
/assign @liggitt 
for API approval

**Comment by ehashman** [comment]
Alpha features job appears to have been broken since July 1: https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-gce-alpha-features?buildId=1410844642957594624

tracked in https://github.com/kubernetes/kubernetes/issues/103440

We should skip that for this and rely on the ubuntu/fedora swap jobs.

/test pull-kubernetes-node-kubelet-swap-ubuntu
/test pull-kubernetes-node-swap-fedora

**Comment by liggitt** [comment]
/approve
for API/config bits

**Pull request merged**
